cal-itp/benefits
# Admin: Add Google SSO for Compiler users

closes #1817 

After merging this PR and ensuring the proper secrets are set up in Terraform, anyone with a `compiler.la` email address will be able to log into the Benefits Admin application. When the user is authenticated, the application will create a user with their Google first name, last name and e-mail address in Benefits. 

## What this PR does
- Uses https://megalus.github.io/django-google-sso/ library for Google single-sign on. Also uses https://github.com/encode/httpx/ for simple HTTP requests, to fetch Google API data.
- Set up Google SSO log in to allow for `compiler.la` domains, if the users email address is in the correct secret (see below)
- Set up Google SSO secrets with Terraform
- Fixes issue where Django JS files were not being allowed on Admin site
- Set user's username as their Google email address, and also save their profile icon URL, first and last name from Google.
- Outside of this PR, the necessary secrets have been set up. These secrets allow us to control exactly which `compiler.la` addresses can log in and get staff or superuser status:

```
GOOGLE_SSO_STAFF_LIST - andy@compiler.la
GOOGLE_SSO_SUPERUSER_LIST - angela@compiler.la,kegan@compiler.la,machiko@compiler.la
```



<img width="559" alt="image" src="https://github.com/cal-itp/benefits/assets/3673236/ec7cfe31-bdb9-4946-a407-b6d2394067fb">
<img width="503" alt="image" src="https://github.com/cal-itp/benefits/assets/3673236/dc54cf0d-89b4-49c9-b38f-1ee5522a1527">

## Done
- [x] Set up Google Cloud SSO new project for Benefits Admin
- [x] Set up OAuth page
- [x] Set up OAuth credentials per environment (Localhost, Dev, Test, Prod). Put in LastPass.
- [x] Set up Terraform with info from above
- [x] Fix test failures
- [x] Host Google image locally
- [x] Make Allowed Domain (`compiler.la, mst.org`) an environment variable (similar to Allowed Hosts - split by comma and filter empties)
- [x] Create username as your Google email address, save user's first name, last name and email address.
- [x] Create Allowable Staff list + env var
- [x] Create Allowable Admin list + env var
- [x] Add spec



## About `django-google-sso`
- Supports Django 5.0 [https://github.com/megalus/django-google-sso/releases](https://github.com/megalus/django-google-sso/releases/tag/v5.0.0)
- Supports Python 3.11 https://megalus.github.io/django-google-sso/#install
- Has a sibling Microsoft and GitHub library: https://megalus.github.io/django-google-sso/multiple/

## How to test
- Engineers: Try to log in via your Compiler email address. Should be able to log in successfully.
- Engineers: Try to log in via a personal Gmail address. Should not be able to log in.
